proposals/004-terminology-for-authentication.md

```diff
@@ -0,0 +1,115 @@
+# Terminology for Authentication
+
+This proposal defines terminology for describing authentication in the proxy. 
+It doesn't propose any changes to any code. 
+It is the first of a number of proposals which will use this terminology.
+
+For the purposes of this proposal we consider TLS and SASL as the only authentication mechanisms we care about.
+In the following, the word "component" is generalising over filters, other plugins, and a proxy virtual cluster as a whole.
+
+## Definitions
+
+### Mutual authentication
+
+Usually 'authentication' is the process by which a client proves their identity to a server. A **mutual authentication mechanism** is one which proves the identity of each party to the other.
+
+### TLS Termination
+
+When the proxy is configured to accept TLS connections from clients it is performing **TLS Termination**, which does not imply mutual authentication. 
+(For the avoidance of doubt, Kroxylicious does not support TLS passthrough; the alternative to TLS Termination is simply using TCP as the application transport).
+
+### TLS client certificate
+
+A **TLS client certificate** is a TLS certificate for the client-side of a TLS Handshake. 
+  For a given client and server pairing, a proxy might have _two_ of these: the Kafka client's TLS client certificate, and the proxy's own TLS client certificate for its connection to the server.
+  
+### Client mutual TLS authentication
+When the proxy is configured to require TLS client certificates from clients and validates these against a set of trusted signing certificates (CA certificate) it is performing **client mutual TLS authentication** ("client mTLS").
+
+### TLS server certificate
+A **TLS server certificate** is a TLS certificate for the server-side of a TLS Handshake. 
+As above, there could be two of these for a given connection through a proxy.
+
+### Server mutual TLS authentication
+
+When the proxy is configured to use a TLS client certificate when making a TLS connection to a server, we will use the term **server mutual TLS authentication_** ("server mTLS").
+
+### SASL Passthrough
+
+a component which forwards a client's `SaslAuthenticate` requests to the server, and conveys the responses back to the client, is performing **SASL Passthrough**.
+
+### Identity preserving
+
+SASL Passthrough is one way to for a proxy to be **identity preserving**, which means that, for all client identities in the virtual cluster, each of those identities will have the same name as the corresponding client identity in the broker.
```

Comment:
Is it worth defining what we/Kafka mean by identity? 

We can't actually ensure that we offer this definition. In the face of opaque tokens we have no way of knowing what the identity in the proxy is or if is the same as the broker holds. So is there a practical difference between `SASL Passthrough` and `Identity Preserving`?

Reply:
I dropped "identity preserving" from the doc. I don't think we really use this term in any of the other documents, so there's no benefit from trying to define it.

FTR I guess there are two orthogonal ideas going on. The definition I wrote required the proxy to have inferred some kind of identity (i.e. assumed passthrough inspection). But the wider idea is that when we're passing through the proxy is not substituting any credentials of its own. I.e. there is only one "domain of users".